### PR TITLE
Checkpoint Usability Initial Changes

### PR DESCRIPTION
**Pull request by ajaysaini725**
Addresses the following in #414
- General docstring cleanup
- The default file format for checkpoints is `.pt` rather than `.tar` except when DeepSpeed is being used in which case the default is `.tar`
- Previous checkpoint file paths are accessible via a trainer property. If files are being uploaded via the `RunDirectoryUploader` then the paths will be storage provider object paths instead.

Also changed saving logic so that checkpoint tarballs are only saved on rank 0 unless deepspeed is being used.

The remaining points in #414 regarding:
1. checkpoint paths being relative to `cwd` instead of the `run_directory`
2. Not requiring the `RunDirectoryUploader` to upload checkpoints (or being more clear about it)
will be addressed by #421 in a later release.

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -439,6 +441,25 @@ def __init__(
     def deepspeed_enabled(self):
         return self.deepspeed_config is not None
 
+    @property
+    def saved_checkpoint_filepaths(self):
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:445]: I'm hesitant about adding this logic into the trainer itself, as a) the trainer should not know about what callbacks are being used, b) all the other checkpointing logic lives in the checkpointer class, and c) it would not be possible for the user to differentiate between getting a local path or the remote path. In addition, these this as written would not work for multi-rank training, since each rank would have it's own checkpoint saver.

Intead, would it work to split this functionality across the checkpoint saver and run directory uploader?

I'm thinking that the checkpoint saver could have a method that returns the paths to all checkpoints. It would perform a `dist.all_gather` to gather any checkpoint files from other ranks, and then return this as a dictionary of {timestamp: list of checkpoint files for that timestamp}. For example:
```python3
def get_saved_checkpoints(self) -> Dict[Timestamp, List[pathlib.Path]]
   """Returns a list of (timestamp, List[filepath]) tuples corresponding to when a checkpoint was saved and all files from all ranks corresponding to this checkpoint. This function performs a distributed reduction to gather all checkpoint files from all ranks"""
   ...
```

Then, could the run directory uploader have a method like this?

```python3
def get_uri_for_uploaded_file(self, local_filepath: Union[pathlib.Path, str]) -> str:
    """"Returns the URI to where ``local_filepath`` was uploaded or would be uploaded. ``local_filepath`` must be in the run directory, or a ValueError would be raised"""
    ...
```

Then, the user would be able to:
```python3

run_directory_uploader = RunDirectoryUploader(...)
trainer = Trainer(callbacks=[run_directory_uploader], checkpoint_folder="checkpoints")
trainer.fit()
for checkpoint_timestamp, checkpoint_filepaths in self.checkpoint_saver.get_saved_checkpoints():
    print("Local filepaths", checkpoint_filepaths)
    print("Uploaded filepaths", [run_directory_uploader.get_uri_for_uploaded_file(x) for x in checkpoint_filepaths])
```

File: composer/callbacks/run_directory_uploader.py [hunk 2]
```diff
@@ -202,6 +211,11 @@ def _trigger_upload(self, logger: Optional[Logger], log_level: Optional[LogLevel
             logger.metric(log_level, {"run_directory/uploaded_files": files_to_be_uploaded})
 
 
+def get_obj_name_for_local_file(object_name_prefix: str, local_filepath: str) -> str:
```
Comment (ravi-mosaicml) [inline composer/callbacks/run_directory_uploader.py:214]:
```suggestion
def _get_obj_name_for_local_file(object_name_prefix: str, local_filepath: str) -> str:
```

Comment (ravi-mosaicml) [inline composer/callbacks/run_directory_uploader.py:214]: Curious why this change?

File: composer/trainer/checkpoint.py [hunk 3]
```diff
@@ -333,17 +339,43 @@ def _get_checkpoint_rng_state(self, checkpoint_rng_state: StateDict) -> Optional
...
 class CheckpointSaver:
-    """Manager for saving state to checkpoint files.
+    """Manager for saving trainer state to checkpoint files.
 
     Args:
         save_folder (str): The path to store checkpoints in.
         interval (Time or str): The amount of time units to wait between checkpoints.
-        compression (str): Compression algorithm to run on checkpoints. Can be `gzip`, `bzip2`,
-            `lzma`, or left blank for no compression.  (default: ``""`` for no compression).
+        compression (str, Optional): Compression algorithm to run on checkpoints. Can be `gzip`, `bzip2`,
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:374]:
```suggestion
        compression (str, optional): Compression algorithm to run on checkpoints. Can be `gzip`, `bzip2`,
```

File: composer/trainer/checkpoint.py [hunk 4]
```diff
@@ -355,20 +387,16 @@ def __init__(self, save_folder: str, interval: Union[Time, str], compression: st
...
-        elif compression == "bzip2":
-            self.write_mode = "w:bz2"
-            self.file_extension = ".tar.bz2"
-        elif compression == "lzma":
-            self.write_mode = "w:xz"
-            self.file_extension = ".tar.lzma"
-        else:
-            raise ValueError(f"Unknown encryption mode: {compression}")
+        self.file_extension, self.write_mode = _format_from_compression(compression=compression)
+        self._save_paths = []
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:391]: I think it'd be fine if this list was public; then no need for the `@property`. Just make sure to have the docstring in the `attributes` secition.

File: composer/trainer/checkpoint.py [hunk 5]
```diff
@@ -394,6 +422,10 @@ def should_checkpoint(self, state: State, event: Event) -> bool:
     def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
         """Save the current state to a a new checkpoint file.
 
+        The default is to save checkpoints in a `.pt` file unless DeepSpeed is being used to train the model
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:425]: Tick marks should probably be double ticks.

File: composer/trainer/checkpoint.py [hunk 6]
```diff
@@ -427,11 +462,17 @@ def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
...
-            checkpoint_archive_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
-            with tarfile.open(checkpoint_archive_filepath, self.write_mode) as tarball:
-                tarball.add(tmpdir, arcname="")  # add files flat to the tarball
-
-            log.info(f'Trainer checkpoint saved to {checkpoint_archive_filepath}')
+            checkpoint_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
+            if not _is_pt_file(checkpoint_filepath):
+                with tarfile.open(checkpoint_filepath, self.write_mode) as tarball:
+                    # add files flat to the tarball with the specified compression
+                    tarball.add(tmpdir, arcname="")
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:469]: Can you set the `arcname` to be something like the checkpointing timestamp and see if that fixes the permission issues we've been seeing?

File: composer/trainer/checkpoint.py (hunk 4 above)
Comment (ravi-mosaicml) [reply composer/trainer/checkpoint.py:391]: Can you also keep track of the `state.timer.get_timestamp()` at which `save_checkpoint` was called?

File: tests/trainer/test_checkpoint.py [hunk 7]
```diff
@@ -114,19 +114,22 @@ def checkpointing_trainer_hparams(composer_trainer_hparams: TrainerHparams) -> T
 def assert_checkpoints_equivalent(hparams_a: TrainerHparams, checkpoint_file_a: str, hparams_b: TrainerHparams,
                                   checkpoint_file_b: str) -> None:
 
+    def _load_checkpoint(checkpoint_dir: str, filename: str):
```
Comment (ravi-mosaicml) [inline tests/trainer/test_checkpoint.py:117]: Can you move this file outside to a top-level function? Not a fan of functions defined in functions...

File: composer/trainer/checkpoint.py [hunk 8]
```diff
@@ -333,17 +339,43 @@ def _get_checkpoint_rng_state(self, checkpoint_rng_state: StateDict) -> Optional
...
+        file_extension = ".tar.gz"
+        write_mode = "w:gz"
+    elif compression == "bzip2":
+        file_extension = ".tar.bz2"
+        write_mode = "w:bz2"
+    elif compression == "lzma":
+        file_extension = ".tar.lzma"
+        write_mode = "w:xz"
+    else:
+        raise ValueError(f"Unknown encryption mode: {compression}")
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:356]:
```suggestion
        raise ValueError(f"Unknown compression mode: {compression}. Valid options are ...")
```

File: composer/callbacks/run_directory_uploader.py (hunk 2 above)
Comment (ajaysaini725) [reply composer/callbacks/run_directory_uploader.py:214]: I wanted to lift the logic for creating the object name by ripping out the run directory ad attaching the prefix out so that it could be imported by the trainer as well

File: composer/trainer/trainer.py (hunk 1 above)
Comment (ajaysaini725) [reply composer/trainer/trainer.py:445]: Yup this sounds good!

File: composer/trainer/checkpoint.py (hunk 6 above)
Comment (ajaysaini725) [reply composer/trainer/checkpoint.py:466]: I tried it didn't :/

File: composer/trainer/checkpoint.py [hunk 9]
```diff
@@ -415,23 +443,39 @@ def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
...
                 with open(composer_states_filepath, 'xb') as f:
                     torch.save(state_dict, f)
 
-            checkpoint_archive_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
-            with tarfile.open(checkpoint_archive_filepath, self.write_mode) as tarball:
-                tarball.add(tmpdir, arcname="")  # add files flat to the tarball
-
-            log.info(f'Trainer checkpoint saved to {checkpoint_archive_filepath}')
+            checkpoint_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
+            if _is_pt_file(checkpoint_filepath):
```
Comment (ajaysaini725) [inline composer/trainer/checkpoint.py:462]: I wanted to avoid nested if statements but intentionally using them here for clarity that there is a unique codepath for `_is_pt_file` being true

**Comment by ajaysaini725** [comment]
Seems like the `test_file_logger` test might need a higher timeout. It failed with a timeout error in Jenkins but I ran it again and the whole pipeline succeeded (see here: https://ci.int.mosaicml.com/blue/organizations/jenkins/scratch%2Fcommand/detail/command/4664/pipeline/24)

File: composer/trainer/checkpoint.py [hunk 10]
```diff
@@ -415,23 +443,39 @@ def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
...
+                    # move the file out of tmpdir to the user-specified location
+                    shutil.move(composer_states_filepath, checkpoint_filepath)
+            else:
+                # using compression, deepspeed, or both
+                with tarfile.open(checkpoint_filepath, self.write_mode) as tarball:
+                    # add files flat to the tarball with the specified compression
+                    tarball.add(tmpdir, arcname="")
+
+            timestamp = state.timer.get_timestamp()
+            paths = dist.all_gather_object(checkpoint_filepath) if is_module_deepspeed(
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:474]: Instead of using a distributed reduction here, I'm wondering if it might be better if I backtrack on my original comment, and instead have each checkpoint saver keep track only of the files it wrote. I'm now realizing that this could be misleading to the user in a multi-node environment, since this function would store paths to files on other nodes. Do you think it would be better if we offloaded this responsibility to the user -- i.e. if you want all checkpoint file paths, perform your own distributed reduction?

Otherwise, I think we should make this extremely clear in the docstring -- perhaps using `.. note::` notation -- and explain that the list index corresponds to the rank on which that the filepath is valid?

File: composer/trainer/checkpoint.py (hunk 9 above)
Comment (ravi-mosaicml) [reply composer/trainer/checkpoint.py:462]: Should the rank zero check be conditioned on whether one is using deepspeed, not whether the checkpoint is ending in  `.pt`? Thinking that if one is using compression and not deepspeed, the only rank zero should make a checkpoint tarball.

**Review by ravi-mosaicml — COMMENTED** [review]
LGTM, though see comments. Can you also add a test for `get_uri_for_uploaded_file`?

File: composer/trainer/checkpoint.py (hunk 10 above)
Comment (ajaysaini725) [reply composer/trainer/checkpoint.py:474]: We should make it as easy as possible for the user to get the checkpoint paths so ideally they don't have to do the reduction themselves. I'll make it extremely clear in the docstring.

File: composer/trainer/checkpoint.py (hunk 9 above)
Comment (ajaysaini725) [reply composer/trainer/checkpoint.py:462]: If someone is using compression you still want to tarball the files though because the original file is written as a .pt file regardless. It is true that with compression + without deepspeed all ranks will be trying to make the tarball but that doesn't do anything bad in terms of performance and makes the code cleaner. 

I thought about making the `else` an `elif is_module_deepspeed(state.model) or dist.get_global_rank() == 0` but that just seemed more confusing from a readability perspective.

I'm not super strongly opinionated about that though so lmk if you want it changed

Comment (ravi-mosaicml) [reply composer/trainer/checkpoint.py:465]: Agree it's not a performance hit, but I think it would be confusing if someone enabled compression and now started to see a bunch of tarballs showing up on non-zero ranks. They may not realize that these can safely be ignored. I know this PR didn't introduce this behavior, but it would be great if we could clean it up here.

File: composer/trainer/checkpoint.py [hunk 11]
```diff
@@ -227,9 +246,9 @@ def _download_checkpoint(self, node_checkpoint_folder: str) -> Tuple[str, Option
             # will detect the process crash and terminate the other ranks
             dist.barrier()
 
-        return composer_checkpoint_filepath, extracted_checkpoint_folder
+        return composer_checkpoint_filepath, extracted_checkpoint_folder, extracted_rank_n
 
-    def _restore_checkpoint(self, state: State, composer_checkpoint_filepath: str,
+    def _restore_checkpoint(self, state: State, composer_checkpoint_filepath: str, extracted_rank_n: bool,
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:251]: extracted_rank_n needs a docstring update

**Review by ravi-mosaicml — APPROVED** [review]
LGTM, though missing one docstring. Thanks for fixing this up!
